### PR TITLE
give better error message when user can't join team due to AR restriction

### PR DESCRIPTION
**Pull request by leem42**

### CODE REVIEW

**Review by brucehoff — CHANGES_REQUESTED** [review]
Please review canAddTeamMember().   There are multiple reasons it can return false.  Your message is specific to one of the reasons.  In the proposed code change there are a number of situations in which the message would not correctly describe the reason.

One approach might be to change canAddTeamMember() to include an object having a boolean and a 'reason' string.  For example see org.sagebionetworks.repo.model.auth.AuthorizationStatus()which does something similar.  Then canAddTeamMember() could return an error message customized for the particular situation causing it to return 'false'.

See TeamManagerImplTest for a variety of reasons that canAddTeamMember() would return false.

**Comment by leem42** [comment]
@brucehoff Thank you for pointing out those issues, looking at the [AuthorizationStatus](https://github.com/Sage-Bionetworks/Synapse-Repository-Services/blob/e57e8fab8e0ee6329526260684cfb4d02db64188/lib/models/src/main/java/org/sagebionetworks/repo/model/auth/AuthorizationStatus.java) class it looks like I could return that object from the canAddTeamMember since it has a method `isAuthorized` and `getMessage`. Would you also agree with returning AuthorizationStatus from that method?

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/TeamManagerImpl.java [hunk 1]
```diff
@@ -123,6 +124,11 @@
 	public static final String USER_HAS_JOINED_TEAM_TEMPLATE = "message/userHasJoinedTeamTemplate.html";
 	public static final String ADMIN_HAS_ADDED_USER_TEMPLATE = "message/teamAdminHasAddedUserTemplate.html";
 	private static final String JOIN_TEAM_CONFIRMATION_MESSAGE_SUBJECT = "New Member Has Joined the Team";
+	private static final String MSG_CANNOT_ADD_TEAM_MEMBER_UNMET_AR = "Cannot add member to team because they have not met all access restrictions. Please remove the pending request and then invite the member again. They will then be prompted to meet the requirement(s) before joining the team.";
+	private static final String MSG_CANNOT_ADD_TEAM_MEMBER_GENERIC = "Cannot add member to team";
+	public static final AuthorizationStatus UNAUTHORIZED_ADD_TEAM_MEMBER_GENERIC =  AuthorizationStatus.accessDenied(TeamManagerImpl.MSG_CANNOT_ADD_TEAM_MEMBER_GENERIC);
+	public static final AuthorizationStatus UNAUTHORIZED_ADD_TEAM_MEMBER_UNMET_AR =  AuthorizationStatus.accessDenied(TeamManagerImpl.MSG_CANNOT_ADD_TEAM_MEMBER_UNMET_AR);
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/TeamManagerImpl.java]: you don't need "TeamManagerImpl." on ll. 129, 130

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/TeamManagerImpl.java [hunk 2]
```diff
@@ -488,30 +494,30 @@ have MEMBERSHIP permission on Team and membership request has been created (but
 	 * @param principalId the ID of the one to be added to the team
 	 * @return
 	 */
-	public boolean canAddTeamMember(UserInfo userInfo, String teamId, UserInfo principalUserInfo, boolean alreadyInTeam) throws NotFoundException {
-		if (userInfo.isAdmin()) return true;
-		if (hasUnmetAccessRequirements(principalUserInfo, teamId)) return false;
+	public AuthorizationStatus canAddTeamMember(UserInfo userInfo, String teamId, UserInfo principalUserInfo, boolean alreadyInTeam) throws NotFoundException {
+		if (userInfo.isAdmin()) return AUTHORIZED_ADD_TEAM_MEMBER;
+		if (hasUnmetAccessRequirements(principalUserInfo, teamId)) return UNAUTHORIZED_ADD_TEAM_MEMBER_UNMET_AR;
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/TeamManagerImpl.java]: 1) Public team and userInfo.equals(principalUserInfo) -> "You can't join the team until you meet the ARs"
2) Private team and the prospective member is calling -> "You can't join the team until you meet the ARs"
3) Private team and the manager is calling AND there's no outstanding request  -> "You can't add them to the team until they've met the ARs and requested to join the team"
4) Private team and the manager is calling AND there's an outstanding request -> UNAUTHORIZED_ADD_TEAM_MEMBER_UNMET_AR

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/TeamManagerImpl.java [hunk 3]
```diff
@@ -488,30 +494,30 @@ have MEMBERSHIP permission on Team and membership request has been created (but
...
 			// if the team is open, I can join
 			Team team = teamDAO.get(teamId);
-			if (team.getCanPublicJoin()!=null && team.getCanPublicJoin()==true) return true;
+			if (team.getCanPublicJoin()!=null && team.getCanPublicJoin()==true) return AUTHORIZED_ADD_TEAM_MEMBER;
 			// if I'm not a team admin and the team is not open, then I need to have an open invitation
-			if (alreadyInTeam) return true;
+			if (alreadyInTeam) return AUTHORIZED_ADD_TEAM_MEMBER;
 			long openInvitationCount = membershipInvitationDAO.getOpenByTeamAndUserCount(Long.parseLong(teamId), Long.parseLong(principalId), now);
-			return openInvitationCount>0L;
+			return openInvitationCount>0L ? AUTHORIZED_ADD_TEAM_MEMBER : UNAUTHORIZED_ADD_TEAM_MEMBER_GENERIC;
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/TeamManagerImpl.java]: "An invitation is required to join the team."

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/TeamManagerImpl.java [hunk 4]
```diff
@@ -488,30 +494,30 @@ have MEMBERSHIP permission on Team and membership request has been created (but
...
 		} else {
 			// the member to be added is someone other than me
-			if (!amTeamAdmin) return false; // can't add somone unless I'm a Team administrator
+			if (!amTeamAdmin) return UNAUTHORIZED_ADD_TEAM_MEMBER_GENERIC; // can't add someone unless I'm a Team administrator
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/TeamManagerImpl.java]: "You must be a team manager to perform this operation."

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/TeamManagerImpl.java [hunk 5]
```diff
@@ -488,30 +494,30 @@ have MEMBERSHIP permission on Team and membership request has been created (but
...
 			// can't add someone unless they are asking to be added
-			if (alreadyInTeam) return true;
+			if (alreadyInTeam) return AUTHORIZED_ADD_TEAM_MEMBER;
 			long openRequestCount = membershipRequestDAO.getOpenByTeamAndRequesterCount(Long.parseLong(teamId), Long.parseLong(principalId), now);
-			return openRequestCount>0L;
+			return openRequestCount>0L  ? AUTHORIZED_ADD_TEAM_MEMBER : UNAUTHORIZED_ADD_TEAM_MEMBER_GENERIC;
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/TeamManagerImpl.java]: "The prospective member must request to join the team."

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/TeamManagerImpl.java [hunk 6]
```diff
@@ -576,7 +582,8 @@ public boolean addMember(UserInfo userInfo, String teamId, UserInfo principalUse
 		String principalId = principalUserInfo.getId().toString();
 		Set<Long> currentMembers = groupMembersDAO.getMemberIdsForUpdate(Long.valueOf(teamId));
 		boolean alreadyInTeam = currentMembers.contains(principalUserInfo.getId());
-		if (!canAddTeamMember(userInfo, teamId, principalUserInfo, alreadyInTeam)) throw new UnauthorizedException("Cannot add member to Team.");
+		AuthorizationStatus canAddTeamMemberStatus = canAddTeamMember(userInfo, teamId, principalUserInfo, alreadyInTeam);
+		if (!canAddTeamMemberStatus.isAuthorized()) throw new UnauthorizedException(canAddTeamMemberStatus.getMessage());
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/TeamManagerImpl.java]: canAddTeamMemberStatus.checkAuthorizationOrElseThrow();

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/team/TeamManagerImplTest.java [hunk 7]
```diff
@@ -515,56 +515,55 @@ public void testCanAddTeamMemberSELF() throws Exception {
 		
 		// I can add myself if I'm an admin on the Team
 		when(mockAuthorizationManager.canAccess(userInfo, TEAM_ID, ObjectType.TEAM, ACCESS_TYPE.TEAM_MEMBERSHIP_UPDATE)).thenReturn(AuthorizationStatus.authorized());
-		assertTrue(teamManagerImpl.canAddTeamMember(userInfo, TEAM_ID, userInfo, false));
+		assertEquals(teamManagerImpl.canAddTeamMember(userInfo, TEAM_ID, userInfo, false), teamManagerImpl.AUTHORIZED_ADD_TEAM_MEMBER);
```
Comment (brucehoff) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/team/TeamManagerImplTest.java]: TeamManagerImpl.AUTHORIZED_ADD_TEAM_MEMBER (18 times)

**Review by brucehoff — CHANGES_REQUESTED** [review]

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/TeamManagerImpl.java [hunk 8]
```diff
@@ -484,41 +490,63 @@ private boolean hasUnmetAccessRequirements(UserInfo memberUserInfo, String teamI
...
 	 * @param userInfo
 	 * @param teamId the ID of the team
-	 * @param principalId the ID of the one to be added to the team
 	 * @return
 	 */
 	public AuthorizationStatus canAddTeamMember(UserInfo userInfo, String teamId, UserInfo principalUserInfo, boolean alreadyInTeam) throws NotFoundException {
+		// admin can always accept membership
 		if (userInfo.isAdmin()) return AUTHORIZED_ADD_TEAM_MEMBER;
-		if (hasUnmetAccessRequirements(principalUserInfo, teamId)) return UNAUTHORIZED_ADD_TEAM_MEMBER_UNMET_AR;
+		if (alreadyInTeam) return AUTHORIZED_ADD_TEAM_MEMBER;
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/TeamManagerImpl.java:525]: What are the implications of returning 'true' when the member is already in the team?

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/TeamManagerImpl.java [hunk 9]
```diff
@@ -484,41 +490,63 @@ private boolean hasUnmetAccessRequirements(UserInfo memberUserInfo, String teamI
...
-			if (alreadyInTeam) return AUTHORIZED_ADD_TEAM_MEMBER;
-			long openRequestCount = membershipRequestDAO.getOpenByTeamAndRequesterCount(Long.parseLong(teamId), Long.parseLong(principalId), now);
-			return openRequestCount>0L  ? AUTHORIZED_ADD_TEAM_MEMBER : UNAUTHORIZED_ADD_TEAM_MEMBER_GENERIC;
+			// check that principalUserInfo has requested to join this team
+			hasInviteOrRequestToJoin = this.checkCanAddOther(teamId, amTeamAdmin, principalId, now);
+		}
+		if (!hasInviteOrRequestToJoin.isAuthorized()) {
+			return hasInviteOrRequestToJoin;
 		}
+		if (hasUnmetAccessRequirements(principalUserInfo, teamId)) {
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/TeamManagerImpl.java:545]: // if principal is other and the caller is allowed to add then 'other' must have requested membership

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/TeamManagerImpl.java [hunk 10]
```diff
@@ -484,41 +490,63 @@ private boolean hasUnmetAccessRequirements(UserInfo memberUserInfo, String teamI
...
 		if (userInfo.isAdmin()) return AUTHORIZED_ADD_TEAM_MEMBER;
-		if (hasUnmetAccessRequirements(principalUserInfo, teamId)) return UNAUTHORIZED_ADD_TEAM_MEMBER_UNMET_AR;
+		if (alreadyInTeam) return AUTHORIZED_ADD_TEAM_MEMBER;
+
 		String principalId = principalUserInfo.getId().toString();
-		boolean principalIsSelf = userInfo.getId().toString().equals(principalId);
+		boolean principalIsSelf = userInfo.equals(principalUserInfo);
 		boolean amTeamAdmin = authorizationManager.canAccess(userInfo, teamId, ObjectType.TEAM, ACCESS_TYPE.TEAM_MEMBERSHIP_UPDATE).isAuthorized();
 		long now = System.currentTimeMillis();
+		AuthorizationStatus hasInviteOrRequestToJoin = null;
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/TeamManagerImpl.java]: rename the variable, e.g. to 'canJoin'

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/TeamManagerImpl.java [hunk 11]
```diff
@@ -484,41 +490,63 @@ private boolean hasUnmetAccessRequirements(UserInfo memberUserInfo, String teamI
...
+		long openRequestCount = membershipRequestDAO.getOpenByTeamAndRequesterCount(Long.parseLong(teamId), Long.parseLong(principalId), now);
+		return openRequestCount>0L  ? AUTHORIZED_ADD_TEAM_MEMBER : UNAUTHORIZED_ADD_TEAM_MEMBER_MUST_HAVE_REQUEST;
+	}
+
+	private AuthorizationStatus checkCanAddSelf(String teamId, boolean amTeamAdmin, String principalId, long now) {
+		// trying to add myself to Team.
+		if (amTeamAdmin) return AUTHORIZED_ADD_TEAM_MEMBER;
+		// if the team is open, I can join
+		Team team = teamDAO.get(teamId);
+		if (team.getCanPublicJoin()!=null && team.getCanPublicJoin()==true) return AUTHORIZED_ADD_TEAM_MEMBER;
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/TeamManagerImpl.java]: use BooleanUtils.isTrue() in apache commons library

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/TeamManagerImpl.java [hunk 12]
```diff
@@ -484,41 +490,63 @@ private boolean hasUnmetAccessRequirements(UserInfo memberUserInfo, String teamI
...
-			long openRequestCount = membershipRequestDAO.getOpenByTeamAndRequesterCount(Long.parseLong(teamId), Long.parseLong(principalId), now);
-			return openRequestCount>0L  ? AUTHORIZED_ADD_TEAM_MEMBER : UNAUTHORIZED_ADD_TEAM_MEMBER_GENERIC;
+			// check that principalUserInfo has requested to join this team
+			hasInviteOrRequestToJoin = this.checkCanAddOther(teamId, amTeamAdmin, principalId, now);
+		}
+		if (!hasInviteOrRequestToJoin.isAuthorized()) {
+			return hasInviteOrRequestToJoin;
 		}
+		if (hasUnmetAccessRequirements(principalUserInfo, teamId)) {
+			if (userInfo.equals(principalUserInfo)) {
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/TeamManagerImpl.java]: use principalIsSelf

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/TeamManagerImpl.java [hunk 13]
```diff
@@ -484,41 +490,63 @@ private boolean hasUnmetAccessRequirements(UserInfo memberUserInfo, String teamI
...
+
+
 	/**
 	 * Either:
 		principalId is self and membership invitation has been extended (and not yet accepted), or
     	principalId is self and have MEMBERSHIP permission on Team, or
     	have MEMBERSHIP permission on Team and membership request has been created (but not yet accepted) for principalId
 	 * @param userInfo
 	 * @param teamId the ID of the team
-	 * @param principalId the ID of the one to be added to the team
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/TeamManagerImpl.java:488]: add the method parameters principalUserInfo, alreadyInTeam

**Review by brucehoff — APPROVED** [review]
Looks great!

**Pull request merged**